### PR TITLE
[2.3][Datasets] Fix filter logic and reuse output buffer 

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

Cherry picking https://github.com/ray-project/ray/pull/32160 into `releases/2.3.0` branch, because this is to fix a new regression introduced in master after 2.2 (affecting correctness of `Dataset.filter()`).
## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
cc @cadedaniel just FYI.

**Comment by cadedaniel** [comment]
Are there any release tests that this will impact? Can we run those before merging to derisk?

**Comment by c21** [comment]
> Are there any release tests that this will impact? Can we run those before merging to derisk?

@cadedaniel - the current Data nightly tests are not impacted by this. We can wait for [today's master nightly test](https://buildkite.com/ray-project/release-tests-branch/builds/1329) to be sure. Given data nightly tests were not failed when initial PR was merged, so it won't be impacted by this fix neither.

The added unit test in this PR help catch the issue in the future.

**Comment by cadedaniel** [comment]
Sounds good. LGTM assuming relevant nightly tests look good.

**Review by cadedaniel — APPROVED** [review]

**Pull request merged**
